### ISSUE TITLE
build_docs swallows errors b/c of changed code in Sphinx

### ISSUE THREAD

**Issue by alexrudy** [astropy/astropy-helpers#288]
When there is an error in the Sphinx build, fails in astropy-helpers' build_docs command fails with 
```
Traceback (most recent call last):
  File "<stdin>", line 39, in <module>
NameError: name 'self' is not defined
Sphinx Documentation subprocess failed with return code 1
```

This is because astropy injects the run method of build_docs into a subprocess, but doesn't inject the command object itself. The command object (``self``) is referenced in [line 190 of sphinx/setup_command.py](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/setup_command.py#L190). The code generated for the subprocess is

```
from __future__ import print_function
from sphinx.setup_command import *

os.chdir(u'/Users/alexrudy/Dropbox/Library/Development/Astronomy/Zeeko/docs')
sys.path.insert(0, u'/Users/alexrudy/Dropbox/Library/Development/Astronomy/Zeeko/build/lib.macosx-10.11-x86_64-2.7')
sys.path.insert(0, '/Users/alexrudy/Dropbox/Library/Development/Astronomy/Zeeko/astropy_helpers')

if not color_terminal():
    nocolor()
if not 3:
    status_stream = StringIO()
else:
    status_stream = sys.stdout
confoverrides = {}
if '':
    confoverrides['project'] = ''
if '':
    confoverrides['version'] = ''
if '':
    confoverrides['release'] = ''
if '':
    confoverrides['today'] = ''
if '':
    confoverrides['copyright'] = ''

app = None
try:
    with docutils_namespace():
        app = Sphinx(u'./docs', u'./docs',
                     u'./build/sphinx/html', u'./build/sphinx/doctrees',
                     'html', confoverrides, status_stream,
                     freshenv=False,
                     warningiserror=False)
        app.build(force_all=False)
        if app.statuscode:
            raise DistutilsExecError(
                'caused by %s builder.' % app.builder.name)
except Exception as exc:
    handle_exception(app, self, exc, sys.stderr)
    if not False:
        raise SystemExit(1)

if False:
    src = app.config.master_doc + app.builder.out_suffix
    dst = app.builder.get_outfilename('index')
    os.symlink(src, dst)

```

I'm not sure what to do about this. Maybe pickle the command object? That seems bad, and swallowing the error is really frustrating.

**Comment by saimn** [comment]
Agreed, this behaviour is very annoying as it hides the real errors (for instance missing dependencies, as in https://github.com/astropy/astropy/issues/5019#issuecomment-275610995). One option could be to create an object that mimics what is needed for [`handle_exception`](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/cmdline.py#L66) (it needs only 3 boolean attributes for pdb, verbosity and traceback).

**Comment by bsipocz** [comment]
@Cadair - this is for you, I knew that I saw it before...

**Comment by bsipocz** [comment]
Fixed by #292.
